app/oauth.go

```diff
@@ -765,17 +765,30 @@ func (a *App) GetAuthorizationCode(w http.ResponseWriter, r *http.Request, servi
 
 	redirectURI := siteURL + "/signup/" + service + "/complete"
 
-	authURL := endpoint + "?response_type=code&client_id=" + clientId + "&redirect_uri=" + url.QueryEscape(redirectURI) + "&state=" + url.QueryEscape(state)
+	authURL, sErr := url.Parse(endpoint)
+	if sErr != nil {
+		return "", model.NewAppError("GetAuthorizationCode.ParseEndpoint", "api.user.get_authorization_code.endpoint_parse.app_error", nil, "failed to parse authentication endpoint URL", http.StatusBadRequest).Wrap(sErr)
+	}
+	queryParams := authURL.Query()
+	queryParams.Set("response_type", "code")
+	queryParams.Set("client_id", clientId)
+	queryParams.Set("redirect_uri", redirectURI)
+	queryParams.Set("state", state)
+
+	authURL.RawQuery = queryParams.Encode()
+	authURLString := authURL.String()
 
+	// Following params aren't being "Set" because URLEncode uses '%20' and QueryEscape uses +,
+	// this change "might" break for some SSO Servers + Mattermost.
```

Comment:
What do you mean by...

> this change "might" break for some SSO Servers + Mattermost.

Reply:
Changed the comment, I meant that if we used `queryParams.Set` on the following parameters, SSO servers expecting `%20` would get `+` instead for encoded space characters.